Dj compact

ywwg opened the pull request:
I need to do a little testing but I think this is good enough for initial commit.

feel free to bikeshed the naming scheme

ywwg committed:
Hercules DJControl Compact: Initial commit, many functions working

ywwg committed:
DJControl Compact: Finish mapping

It's possible to send different light information when Shift is held.
For now we just send the same lights either way.

ywwg committed:
DJControl Compact: tabs to spaces

Be-ing commented on res/controllers/DJControl-Compact-scripts.js:
```diff
@@ -0,0 +1,113 @@
+}
+
+HercDJCompact.init = function(id) {
+    scratch = false;
```
Should this be an array to keep track of it for each deck?

Be-ing replied:
Nevermind, I see that this device has a single "Scratch" button for both decks.

Be-ing commented on res/controllers/DJControl-Compact-scripts.js:
```diff
@@ -0,0 +1,113 @@
+        value = value - 0x80;
+    }
+    if (scratch) {
+        if (value != 0) {
```
Please use `!==` and `===` rather than `!=` and `==` to avoid unintended oddities with JS type conversions

Be-ing commented:
Do you plan on documenting this now or when you finish the mapping? There's already a wiki page started for a different, unfinished mapping: http://mixxx.org/wiki/doku.php/hercules_djcontrol_compact

Be-ing commented on res/controllers/DJControl-Compact-scripts.js:
```diff
@@ -0,0 +1,113 @@
+            }
+        }
+
+        if(engine.getValue(input.group, "scratch2_enable")) {
```
IMO the code would be easier to understand if this was moved up above line 52. I wasn't clear what lines 54-56 did until I read this far.

Be-ing commented on res/controllers/DJControl-Compact-scripts.js:
```diff
@@ -0,0 +1,113 @@
+// Mode button does give out midi, but the selection itself is done inside the
+// controller -- no control over which mode is selected for instance.
```
By this do you mean the pads send different MIDI messages when different modes are selected?

ywwg replied:
to select pad modes, you hold the mod button and then press one of the pads.  That functionality is in the fimeware of the controller and can't be changed.  And then yes, the pads send different midi signals depending on what modes are selected.  All buttons also send different messages when shift is held, so I don't even have to program the shift button in js!  (I'll be filling out all this information in comments in a subsequent commit)

Be-ing commented on res/controllers/DJControl-Compact-scripts.js:
```diff
@@ -0,0 +1,113 @@
+    scratch_timer_on[n] = false;
+}
+
+HercDJCompact.pitch = function (group, control, value, status) {
```
I'm confused what this is for considering there's no pitch fader on this device.

ywwg replied:
shift + jog wheel sends a different midi signal, and this is hooked up to that.

Be-ing replied:
It would be helpful to add a comment explaining that.

Be-ing commented:
Looks good so far other than the minor issues pointed out above. There's not much more to comment on without documentation.

Be-ing commented:
Before @sblaisot has to jump in to say it :P
Update the Windows uninstaller to include the new files: https://mixxx.org/wiki/doku.php/contributing_mappings#windows_installer_update

ywwg commented:
> Do you plan on documenting this now or when you finish the mapping? There's already a wiki page started for a different, unfinished mapping: http://mixxx.org/wiki/doku.php/hercules_djcontrol_compact

I was given this controller by hercules for the purpose of making an official mapping, so yes I'll be documenting it.  As I stated this is just an initial commit.

sblaisot commented:
please ad file removal from windows uninstaller
https://mixxx.org/wiki/doku.php/contributing_mappings#windows_installer_update

Be-ing commented on res/controllers/DJControl-Compact-scripts.js:
```diff
@@ -0,0 +1,113 @@
+            "HercDJCompact.OnRecordingStatusChange");
+
+    // Tell controller to send midi to update knob and slider positions.
+    midi.sendShortMsg(0xB0, 0x7F, 0x7F);
```
Cool! I didn't see anything about this in the P32 MIDI documentation, but it works on that too! :)

ywwg replied:
yeah it's a great feature.  (It was doc'ed in the pdf on hercules' site).  I'm really impressed with the midi programming on this controller, a lot of convenience features.

Be-ing replied:
I tried this with my Electrix Tweaker just to see if it would work but it didn't. Is this a Hercules thing? I think I recall looking at one other mapping that had such a feature, but I don't remember what mapping it was or what MIDI signal was sent.

ywwg replied:
my vestax VCI400 just does it on connect, for instance, no midi trigger required.  It's certainly not any sort of standard, so I'd guess that someone at Hercules thought it up and they reuse firmwares on multiple devices.

ywwg committed:
Merge branch 'master' into dj-compact

ywwg committed:
DJ Compact: last fixes, correct filenames

ywwg commented:
notes addressed

Be-ing commented:
Hm, considering this controller lacks a high EQ knob, would it be helpful to (optionally?) map the mid EQ knob to the HPF (QuickEffect parameter3) or QuickEffect superknob?

ywwg commented:
I am writing a framework for creating "controller preferences," which will allow this type of setting to be changed in the UI.  For now I'll add a comment that users can remap the knob with the midi wizard (rather than adding javascript), or maybe even put some comments in the xml about how to do this.  

When controller prefs comes in, I'll make a proper set of preferences for the knob.

It is a little silly that it's set to "mid" :/.

ywwg committed:
DJControl Compact: Give the user some help for remapping the mid knob

ywwg committed:
english

Be-ing commented:
> I am writing a framework for creating "controller preferences," which will allow this type of setting to be changed in the UI.

Awesome!

> For now I'll add a comment that users can remap the knob with the midi wizard (rather than adding javascript), or maybe even put some comments in the xml about how to do this. 

Sounds good.

Be-ing commented on res/controllers/Hercules DJControl Compact.midi.xml:
```diff
@@ -12,6 +14,28 @@
+            </control>
+            <control>
+                <group>[EqualizerRack1_[Channel2]_Effect1]</group>
+                <!-- Set the key on the next line to "parameter3" to make the
```
Would it be better to use the <description> element for this? That shows up in the mapping GUI.

ywwg replied:
the description tells them to edit the XML, but I don't think that's the place to go into this detail.  This comment is right at the top of the file so if someone does get as far as editing the XML, they'll find it (I hope)

Be-ing replied:
Alright, fine to leave it as is then

ywwg commented:
lgty?

Be-ing commented on res/controllers/Hercules DJControl Compact.midi.xml:
```diff
@@ -0,0 +1,1219 @@
+        <name>Hercules DJControl Compact</name>
+        <author>Owen Williams</author>
+        <description>Controller mapping for Hercules DJControl Compact.  Users
+who want the labelled "mid" knob to adjust high eq instead may use the midi
```
spelling: "labeled"

Be-ing commented on res/controllers/Hercules DJControl Compact.midi.xml:
```diff
@@ -0,0 +1,1219 @@
+        <author>Owen Williams</author>
+        <description>Controller mapping for Hercules DJControl Compact.  Users
+who want the labelled "mid" knob to adjust high eq instead may use the midi
+wizard to remap it, or edit the controller xml file directly to change this
```
Capitalize MIDI and XML?

Be-ing commented:
JSHint warnings:
res/controllers/Hercules-DJControl-Compact-scripts.js: line 4, col 17, Weird construction. Is 'new' necessary?
res/controllers/Hercules-DJControl-Compact-scripts.js: line 6, col 1, Missing '()' invoking a constructor.
res/controllers/Hercules-DJControl-Compact-scripts.js: line 6, col 2, Missing semicolon.
res/controllers/Hercules-DJControl-Compact-scripts.js: line 52, col 19, Use '!==' to compare with '0'.
res/controllers/Hercules-DJControl-Compact-scripts.js: line 72, col 19, Use '!==' to compare with '0'.
res/controllers/Hercules-DJControl-Compact-scripts.js: line 83, col 2, Missing semicolon.

Be-ing commented:
I'm guessing the beatlooproll_X_activate mappings are for loop mode with shift pressed. Is that correct? Could you document the mapping on http://mixxx.org/wiki/doku.php/hercules_djcontrol_compact ? I think 2x2 tables describing the pad functionality in each mode would be helpful.

Be-ing commented:
> I am writing a framework for creating "controller preferences," which will allow this type of setting to be changed in the UI.

A bit off topic, but how do you intend to implement this? I think passing a QScriptValue containing all the options to the script's init function could work well.

ywwg commented:
wiki will be updated after this is in.

RJ and I have been collaborating on a design for controller preferences, it's similar to what you describe.  Once it's settled I'll post the design doc to the mailing list.

ywwg committed:
DJControl Compact: last notes

ywwg commented:
ping

Be-ing commented:
As I said before, there isn't much to comment on without having documentation to reference. I can't tell if it looks good unless I wen through the MIDI documentation for this device and compared it to the XML.

ywwg commented:
you're not expected to go through every button in/out and confirm that it's correct, this review is mostly for the style of the javascript functions.  I've gone through all the buttons and functions I've mapped and made sure the controller functions nicely.

ywwg committed:
Merge branch 'master' into dj-compact

ywwg committed:
DJControl: Add hotcue clear buttons

ywwg committed:
DJControl Compact: fix autodj binding

ywwg commented:
updated documentation. Most functions are the same as manual (linked on wiki), differences are noted.  http://mixxx.org/wiki/doku.php/djcontrol_compact

Be-ing commented:
Thanks for documenting it. I added pictures of the device to the wiki from [the other wiki page](http://mixxx.org/wiki/doku.php/hercules_djcontrol_compact), reorganized the page, and made some edits. There is so little information to convey that I think referring users to Hercules' manual is more of an annoyance, so I put the little bit of information from there onto the wiki. Please confirm that it's all correct.

I like the way you've mapped the pads in loop mode. That is more flexible than the way described in Hercules' manual.

I think it would be a better use of the sampler buttons to map the left side to samplers 1-4 and right side to samplers 5-8.

It doesn't look like you have mapped shift+play to anything. I find it useful to map this to `start_stop` (go to beginning of track and stop).

ywwg commented:
There's a word or two missing in your edit of the jog wheel instructions: "Enable keylock on a deck by clicking the if you do not want the pitch to change when adjusting the tempo."

I'll let you know when I've addressed the notes.

ywwg commented:
Mixxx only has 4 samplers on startup, so specifying samplers 5-8 causes a warning to appear

ywwg commented:
Do any of our skins support more than four samplers? It doesn't look like either Latenite nor Deere do

ywwg committed:
Add shift-play behavior
Fix jog wheels while playing
Change samplers, although those samplers don't exist

Be-ing commented:
> Mixxx only has 4 samplers on startup, so specifying samplers 5-8 causes a warning to appear

Create more by setting [Master],num_samplers to 8 in the script's init() function.

> Do any of our skins support more than four samplers? It doesn't look like either Latenite nor Deere do

Yes, Deere does by flipping through the multiple pages of samplers available. It's clunky, but it's there.

ywwg committed:
DJ Compact: set sampler count to 8

ywwg commented:
> Create more by setting [Master],num_samplers to 8 in the script's init() function.

done

Be-ing commented:
LGTM :+1:
